src/many-migration/src/lib.rs

```diff
@@ -378,101 +139,300 @@ impl<'a, T, E> InnerMigration<'a, T, E> {
-                    "Migration {} is not of type `Regular`, skipping",
-                    self.name()
-                );
+                trace!("Skipping migration {}", self.name());
                 Ok(())
             }
         }
     }
 
     /// This function gets executed when the storage block height >= the migration block height
-    pub fn update(&self, storage: &mut T) -> Result<(), E> {
+    fn update(&self, storage: &mut T) -> Result<(), E> {
         match &self.r#type {
             MigrationType::Regular(migration) => (migration.update_fn)(storage),
+            MigrationType::Hotfix(_) => Ok(()),
             _ => {
-                tracing::trace!(
-                    "Migration {} is not of type `Regular`, skipping",
-                    self.name()
-                );
+                trace!("Skipping migration {}; not of type `Regular`", self.name());
                 Ok(())
             }
         }
     }
 
     /// This function gets executed when the storage block height == the migration block height
-    pub fn hotfix<'b>(&'b self, b: &'b [u8]) -> Option<Vec<u8>> {
+    fn hotfix<'b>(&'b self, b: &'b [u8]) -> Option<Vec<u8>> {
         match &self.r#type {
+            MigrationType::Regular(_) => None,
             MigrationType::Hotfix(migration) => (migration.hotfix_fn)(b),
             _ => {
-                tracing::trace!(
-                    "Migration {} is not of type `Hotfix`, skipping",
-                    self.name()
-                );
+                trace!("Skipping migration {}; not of type `Hotfix`", self.name());
                 None
             }
         }
     }
 }
 
-#[derive(Deserialize)]
-struct MigrationJson<'a> {
-    r#type: &'a str,
+#[derive(Debug)]
+pub struct Migration<'a, T, E> {
+    migration: &'a InnerMigration<T, E>,
+
+    /// The metadata used during creation of this migration.
+    metadata: Metadata,
+
+    /// Whether the migration is enabled (will initialize, update, etc).
+    enabled: bool,
+
+    /// Whether the block height has been reached.
+    active: bool,
+}
+
+impl<'a, T, E> fmt::Display for Migration<'a, T, E> {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_fmt(format_args!(
+            "{}, Metadata: \"{:?}\", Status: \"{}\"",
+            self.migration,
+            self.metadata(),
+            self.is_enabled()
+        ))
+    }
+}
+
+impl<'a, T, E> Migration<'a, T, E> {
+    fn new(migration: &'a InnerMigration<T, E>, metadata: Metadata) -> Self {
+        let enabled = !metadata.disabled;
+        Self {
+            migration,
+            metadata,
+            enabled,
+            active: false,
+        }
+    }
+
+    /// Check the height and call the inner migration's methods.
+    pub fn apply_height(&mut self, storage: &mut T, block_height: u64) -> Result<(), E> {
+        if self.is_enabled() {
+            if block_height > self.metadata.block_height && !self.active {
```

Comment:
`==` instead of `>`?

Reply:
How did that past the tests :/ Need to test more.